### PR TITLE
tracer/remote-config: Subscribe to dynamic instrumentation configs via remote config

### PR DESCRIPTION
**Pull request by grantseltzer**
### What does this PR do?

This adds a configuration option (DD_DYNAMIC_INSTRUMENTATION_ENABLED) which will tell the tracer to subscribe to dynamic instrumentation configurations via remote config.

It also has a go.mod change to use a recent const I added in the agent's `remoteconfig/state` package. I imagine we want to upgrade to the next release once it's available before merging.

### Motivation

In order to enable Go Dynamic Instrumentation, users will have to be using the Go tracer. This eases the challenge of having Dynamic Instrumentation track what Go services are available to instrument. Since DI needs remote config, it makes sense to have this live in the go tracer. What the DI team is working on currently, is changing the RC backend, and adding logic to the datadog-agent to re-route the returned configurations to the system-probe (where Go-DI will live) instead of going back to the Go tracer. 

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

For Datadog employees:

- [ ] If this PR touches code that handles credentials of any kind, such as Datadog API keys, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

### CODE REVIEW

File: go.mod
Comment (darccio) [inline go.mod:10]: @grantseltzer When is it planned to release this code as part of an agent release? I see it's some kind of dev version.

File: ddtrace/tracer/remote_config.go
Comment (felixge) [inline ddtrace/tracer/remote_config.go:174]: ⚠️ I don't know when/why `RegisterProduct` might fail, but I suspect we might still want to try to register `ProductAPMTracing` below when that happens?

File: go.mod
Comment (grantseltzer) [reply go.mod:10]: I agree we should wait for an actual release, i'll talk to the relevant team today to figure out when that's planned for.

File: ddtrace/tracer/remote_config.go
Comment (cimi) [inline ddtrace/tracer/remote_config.go:188]: Nit: we could just `return remoteconfig.Subscribe(`

File: ddtrace/tracer/remote_config.go
Comment (cimi) [reply ddtrace/tracer/remote_config.go:174]: Agree, this shouldn't affect anything else. 

If registering our product fails we should just log an error and let the program continue.

File: ddtrace/tracer/option.go
Comment (cimi) [reply ddtrace/tracer/option.go:266]: I don't think we need to mention remote configuration in the docs, it's an implementation detail, but don't feel strongly about it 🙂

File: go.mod
Comment (cimi) [reply go.mod:10]: :+1: We're not in a rush with this PR, we can wait for them to release

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go]: That sounds right to me. What do you think of the change I made here? The error return from `startRemoteConfig` get's logged.

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go]: See other comment from Felix about this code, modified accordingly

File: ddtrace/tracer/remote_config.go
Comment (cimi) [inline ddtrace/tracer/remote_config.go:188]: To make sure we're keeping the same behaviour from before, we should return `apmTracingError` and only log `dynamicInstrumentationError`.

I'm happy to change this if we understand what it means to return error from this function 🙂 (or when / if registering a product can fail).

Comment (cimi) [reply ddtrace/tracer/remote_config.go:188]: Sorry, missed your other comment.

If the error is just logged then this is fine :+1:

File: go.mod
Comment (grantseltzer) [reply go.mod:10]: Based on the schedule previous releases it looks like it will be at least a few weeks before the actual release is cut (it's on rc-1 right now). I've changed the code to use the actual string instead of the constant for now, I can go in and use the constant later on.

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go:95]: This is on purpose, it's hooked via bpf

File: ddtrace/tracer/tracer.go
Comment (cimi) [inline ddtrace/tracer/tracer.go:166]: Let's remove this log or make it debug - it's not helpful for us right now and it might be a surprise for other users of RC.

We will review all logs we emit from Go DI before releasing to customers, we need to be sure that we have all the information needed to troubleshoot customer issues.

Typo - succesffully

File: ddtrace/tracer/remote_config.go
Comment (cimi) [inline ddtrace/tracer/remote_config.go:88]: What effect does returning `nil` here have?

If it doesn't acknowledge the config, it means the agent will keep sending it to the tracer and I'm not sure if it impacts the backend.

We should check with RC to see what we should return here.

File: ddtrace/tracer/remote_config.go
Comment (cimi) [inline ddtrace/tracer/remote_config.go:82]: This should be a debug log.

File: go.mod
Comment (cimi) [reply go.mod:10]: Since we removed the imported constant we should also remove the dependency here

File: go.sum
Comment (cimi) [inline go.sum:632]: Let's remove?

File: internal/remoteconfig/remoteconfig.go
Comment (cimi) [inline internal/remoteconfig/remoteconfig.go:237]: Let's change the other two logs here to `Debug` and remove the error log for empty body.

File: ddtrace/tracer/remote_config.go
Comment (cimi) [reply ddtrace/tracer/remote_config.go]: I think it's fine now - we could ignore the DI error and just return the APM result to keep the previous behaviour exactly as it was but returning error when either fail is also fine; to get a DI error users would need to set the env var.

File: ddtrace/tracer/tracer.go
Comment (grantseltzer) [reply ddtrace/tracer/tracer.go:166]: i'll remove for now, I agree that we should review all logs before we release.

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go:88]: Thats a good question. My understanding was that it would be set to an unknown state (state.ApplyStateUnknown) but parsing through the code now I don't see that. I'll look into it more closely on Monday, but returning `state.ApplyStateUnacknowledged` seems to makes sense.

File: ddtrace/tracer/remote_config.go
Comment (cimi) [inline ddtrace/tracer/remote_config.go:85]: This is the correct state to set for the config from the tracer's point of view - we don't know here if the BPF instrumentation was applied correctly or not - but I'm not sure what effect it has on the RC backend.

AFAIK, if configs are acknowledged they are not re-sent by the backend, but I'm not 100% sure. We should check with the RC team, I'll add this to the RFC.

**Review by cimi — APPROVED** [review]
:+1: Thanks!

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [inline ddtrace/tracer/remote_config.go:98]: @felixge I'm curious for your thoughts on this. `golint-ci` is complaining about these unused parameters. I can rename them each `_` and instrumentation seems to work just fine (can still get the parameter values) but perhaps there's other implications i'm unaware of. Renaming them removes entries for parameters in DWARF (though the bpf code for extracting them isn't going through DWARF). 

Is it possible in other architectures or in the future they wouldn't be passed at all and it'd be instrumentation? Would you recommend just overwriting the linter or changing the names to `_` ?

Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go:98]: Disassembly of the call to passFullConfiguration when the params are renamed to `_`:

```
  7dd7a8:       f9403be0        ldr     x0, [sp, #112]
  7dd7ac:       f94023e1        ldr     x1, [sp, #64]
  7dd7b0:       94000038        bl      7dd890 <gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.passFullConfiguration>
 ```
 
I'm going to go ahead and make the change but still curious on your thoughts.

Comment (felixge) [reply ddtrace/tracer/remote_config.go]: Interesting question. I think the behavior you're relying on here is outside of the [language spec](https://go.dev/ref/spec), in particular:

* `//go:noinline` is not part of the language spec, and might break in the future.
* The calling convention for invoking passFullConfiguration() is [ABIInternal](https://go.googlesource.com/go/+/refs/heads/master/src/cmd/compile/abi-internal.md) which "is unstable and will change between Go versions"
* The reliance on the compiler not eliminating the call to `passFullConfiguration()` entirely.
* The reliance on the compiler to pass arguments when they are not used in the function body.
* The reliance on the compiler to pass arguments when they can't be used in the function body when they are being unnamed.

If I was you, I wouldn't be too worried about the last thing, at least not more worried than about the other things. The only way to make this stuff robust is to cover it with tests and run them against future versions of Go before they are being released.

**Comment by grantseltzer** [comment]
@felixge I fixed the broken test. Output was a little hidden but it was because I had multiple `%w` in a call to `fmt.Errorf`

File: ddtrace/tracer/remote_config.go
Comment (katiehockman) [inline ddtrace/tracer/remote_config.go:90]: need to finish the sentence?

**Review by katiehockman — APPROVED** [review]

File: ddtrace/tracer/remote_config.go
Comment (grantseltzer) [reply ddtrace/tracer/remote_config.go:90]: Thanks for catching, the comment was out of date anyway

**Pull request merged**
